swarm/src/behaviour.rs

```diff
@@ -31,26 +31,116 @@ use std::{task::Context, task::Poll};
 type THandlerInEvent<THandler> =
     <<THandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::InEvent;
 
-/// A behaviour for the network. Allows customizing the swarm.
+/// A [`NetworkBehaviour`] defines the behaviour of the local node on the network.
 ///
-/// This trait has been designed to be composable. Multiple implementations can be combined into
-/// one that handles all the behaviours at once.
+/// In contrast to [`Transport`](libp2p_core::Transport) which defines **how** to send bytes on the
+/// network, [`NetworkBehaviour`] defines **what** bytes to send and **to whom**.
 ///
-/// # Deriving `NetworkBehaviour`
+/// Each protocol (e.g. `libp2p-ping`, `libp2p-identify` or `libp2p-kad`) implements
+/// [`NetworkBehaviour`]. Multiple implementations of [`NetworkBehaviour`] can be composed into a
+/// hierarchy of [`NetworkBehaviour`]s where parent implementations delegate to child
+/// implementations. Finally the root of the [`NetworkBehaviour`] hierarchy is passed to
+/// [`Swarm`](crate::Swarm) where it can then control the behaviour of the local node on a libp2p
+/// network.
 ///
-/// Crate users can implement this trait by using the the `#[derive(NetworkBehaviour)]`
+/// # Hierarchy of [`NetworkBehaviour`]
+///
+/// To compose multiple [`NetworkBehaviour`] implementations into a single [`NetworkBehaviour`]
+/// implementation, potentially building a multi-level hierarchy of [`NetworkBehaviour`]s, one can
+/// use one of the [`NetworkBehaviour`] combinators, and/or use the [`NetworkBehaviour`] derive
+/// macro.
+//G/
+/// ## Combinators
+///
+/// [`NetworkBehaviour`] combinators wrap one or more [`NetworkBehaviour`] implementations and
+/// implement [`NetworkBehaviour`] themselves. Example is the [`Toggle`](crate::toggle::Toggle)
+/// [`NetworkBehaviour`].
+///
+/// ``` rust
+/// # use libp2p_swarm::DummyBehaviour;
+/// # use libp2p_swarm::toggle::Toggle;
+/// let my_behaviour = DummyBehaviour::default();
+/// let my_toggled_behaviour = Toggle::from(Some(my_behaviour));
+/// ```
+///
+/// ## Derive Macro
+///
+/// One can derive [`NetworkBehaviour`] for a custom `struct` via the `#[derive(NetworkBehaviour)]`
 /// proc macro re-exported by the `libp2p` crate. The macro generates a delegating `trait`
 /// implementation for the `struct`, which delegates method calls to all struct members.
 ///
-/// Struct members that don't implement [`NetworkBehaviour`] must be annotated with `#[behaviour(ignore)]`.
-///
 /// By default the derive sets the [`NetworkBehaviour::OutEvent`] as `()` but this can be overridden
```

Comment:
With the change to `event_process = false`, this might actually be a bad default now.

Reply:
In theory, we could have the derive auto-generate an enum but having that one be "invisible" could be very strange. Thoughts?

Reply:
> With the change to event_process = false, this might actually be a bad default now.

Agreed.

> In theory, we could have the derive auto-generate an enum but having that one be "invisible" could be very strange. 

I am not in favor of auto-generating the events enum. Instead, I would suggest making `event_process` a mandatory field to be set by the user.

Reply:
Let's continue the discussion in https://github.com/libp2p/rust-libp2p/issues/2373.